edxlearndot/learndot.py

```diff
@@ -363,14 +399,132 @@ def set_enrolment_status(self, enrolment_id, status):
+                        "Learndot enrolment was logged as set to %s at %s, so skipping update.",
+                        enrolment_status.status,
+                        enrolment_status.updated_at
+                    )
+                    return
+            except EnrolmentStatusLog.DoesNotExist:
+                pass
+
         response = requests.post(
             self.get_enrolment_management_url_template().format(enrolment_id=enrolment_id),
             headers=self.get_api_request_headers(),
             json={"status": status}
         )
+
         try:
             response.raise_for_status()
         except requests.exceptions.HTTPError as e:
             msg = "Error trying to set status of enrolment {} to {}: {}".format(enrolment_id, status, e)
             log.error(msg)
             raise LearndotAPIException(msg)
+
+        try:
+            enrolment_status_log, _created = EnrolmentStatusLog.objects.get_or_create(
+                learndot_enrolment_id=enrolment_id
+            )
+            enrolment_status_log.status = status
+            enrolment_status_log.save()
+            log.info(
+                "Recorded status of enrolment %s as %s at %s",
+                enrolment_status_log.learndot_enrolment_id,
+                enrolment_status_log.status,
+                enrolment_status_log.updated_at
+            )
+        except (IntegrityError, MultipleObjectsReturned) as e:
+            log.error("Error recording enrolment status update: %s", e)
+
+    def set_enrolment_status_to_passed(self, enrolment_id, unconditional=False):
+        """
+        Sets the status of a Learndot enrollment record to PASSED.
+        Arguments:
+            enrolment_id (int): the numeric Learndot enrollment ID
+            unconditional (bool): whether to unconditionally send the new
+                                  status to Learndot, instead of checking
+                                  EnrolmentStatusLog records to see if it
+                                  has already been sent.
+        Returns:
+            None
+        Raises:
+            LearndotAPIException: if the requested status is invalid, or
+                Requests throws anything.
+        """
+        self.set_enrolment_status(enrolment_id, EnrolmentStatus.PASSED, unconditional)
+
+
+class LearndotAPIClientMock(LearndotAPIClientBase):
+    """
+    Mock client for tests.
+    """
+    DOES_NOT_EXIST_ID = 1
+
+    enrolments = {}
+
+    def get_api_key(self):
+        return "bababababababababababa"
+
+    def get_api_base_url(self):
+        return "https://localhost/learndot/v2/api"
+
+    def get_contact_id(self, user):
+        """
+        Mock implementation just returns the user ID as the contact ID.
+
+        If the user is None, or the user ID is self.DOES_NOT_EXIST_ID,
+        the mock contact does not exist.
+        """
+
+        if user is None or user.id == self.DOES_NOT_EXIST_ID:
+            return None
+
+        return user.id
```

Comment:
So that the actual `learndot.LearndotAPIClient.get_contact_id` code is tested, could this code stub out the API call and [mock the response](https://pypi.python.org/pypi/responses), instead of just bypassing the code entirely?

Would also like to see tests for the caching behaviour added by this change.

Ditto for the other methods below..  As the coverage currently stands, even the `LearndotAPIClientMock.get_api_key` and `LearndotAPIClientMock.get_api_base_url` methods aren't called.